File: samples/grpc-reactive/src/main/java/org/springframework/grpc/sample/GrpcServerApplication.java
package org.springframework.grpc.sample;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.grpc.server.exception.GrpcExceptionHandler;

import io.grpc.Status;

@SpringBootApplication
public class GrpcServerApplication {

	private static Log log = LogFactory.getLog(GrpcServerApplication.class);

	public static void main(String[] args) {
		SpringApplication.run(GrpcServerApplication.class, args);
	}

	@Bean
	GrpcExceptionHandler grpcExceptionHandler() {
		return ex -> {
			if (ex instanceof IllegalArgumentException) {
				log.error("Error in grpc exception", ex);
				return Status.INVALID_ARGUMENT.withDescription(ex.getMessage()).asException();
			}
			return Status.INTERNAL.withCause(ex).withDescription(ex.getMessage()).asException();
		};
	}

}

File: samples/grpc-server-kotlin/src/main/kotlin/org/springframework/grpc/sample/GrpcServerApplication.kt
package org.springframework.grpc.sample

import io.grpc.Status
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.context.annotation.Bean
import org.springframework.grpc.server.GlobalServerInterceptor
import org.springframework.grpc.server.exception.GrpcExceptionHandler
import org.springframework.grpc.server.security.AuthenticationProcessInterceptor
import org.springframework.grpc.server.security.GrpcSecurity
import org.springframework.security.config.Customizer.withDefaults

@SpringBootApplication
open class GrpcServerApplication {

    @Bean
    open fun globalInterceptor(): GrpcExceptionHandler = GrpcExceptionHandler { exception ->
        when (exception) {
            is IllegalArgumentException -> Status.INVALID_ARGUMENT.withDescription(exception.message).asException();
            else -> null
        }
    }

    @Bean
    @GlobalServerInterceptor
    @Throws(Exception::class)
    open fun jwtSecurityFilterChain(grpc: GrpcSecurity): AuthenticationProcessInterceptor {
        return grpc
            .authorizeRequests { requests ->
                requests.methods("Simple/StreamHello")
                    .hasAuthority("SCOPE_profile")

                requests.methods("Simple/SayHello")
                    .authenticated()
                    .methods("grpc.*/*")
                    .permitAll()
                requests.allRequests()
                    .denyAll()
            }
            .oauth2ResourceServer { resourceServer ->
                resourceServer.jwt(withDefaults())
            }
            .build()
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            runApplication<GrpcServerApplication>(*args)
        }
    }
}


